Issue 37 kakao pay impl 리팩터링

Bue-von-hon opened the pull request:
### 작업사항
* 기존에 가격을 받아 결제요청하는 로직을 주문 클래스를 받아 요청하도록 변경했습니다
* 주문 객체에 담겨있는 내용을 바탕으로 결제 응답을 받을 수 있도록 수정했습니다
* 결제 응답 객체의 http 응답 코드를 비교하여 승인되었는지 확인합니다
* Order 클래스의 id 타입을 long -> Long 수정(null이 가능할 수 있게 변경)

Bue-von-hon committed:
processPay 메소드 간결화

Bue-von-hon committed:
id가 null이 가능하도록 레퍼 클래스로 변경

Bue-von-hon committed:
Order를 받아 결제하도록 변경

Bue-von-hon committed:
결제 응답을 받기 위한 클래스 생성

Bue-von-hon committed:
할인된 주문을 리턴하도록 변경

Bue-von-hon committed:
가격이 아닌 주문 클래스를 받아 처리하도록 수정

Bue-von-hon committed:
가격이 아닌 주문 클래스를 받아 처리하도록 수정

Bue-von-hon committed:
깨진 테스트 수정

Taewoongjung commented on src/main/java/app/foodylab/adaptor/payment/KakaoPayResp.java:
```diff
@@ -0,0 +1,17 @@
+
+@Getter
+@RequiredArgsConstructor
+public class KakaoPayResp {
```
Resp 보다 Response로 풀네임이 의미가 더 명확히 전달되어 잘 읽힐 것 같습니다.

Bue-von-hon replied:
반영했습니다!
[e61d912](https://github.com/f-lab-edu/foody-lab/pull/43/commits/e61d9128f6bf8766e5feaa5198ea60e9d8e05b23)

Bue-von-hon committed:
resp -> response 풀네임으로 변경

Taewoongjung commented on src/main/java/app/foodylab/application/payment/service/PayService.java:
```diff
@@ -30,7 +30,7 @@ final void init() {
         PayMethod method = payMethodMap.get(payMethod);
-        long price = discountPriceCalculateService.getPrice(order, coupons);
-        method.pay(price);
+        Order discountOrder = discountPriceCalculateService.getPrice(order, coupons);
```
저희 쿠폰 계산은 Order 도메인에서 하기로 했지 않았나요??

Bue-von-hon replied:
사실 이 부분 구현은 다음과 같은 이유로 고민중입니다.

주문 클래스가 쿠폰 클래스를 "**포함**"해야 할까?
만약 주문 클래스에게 "최종 금액"을 질의해야 한다면, 쿠폰 클래스도 들고 있어야합니다.
책임이 너무 몰리는게 아닌지 고민입니다.

좋은 아이디어 있으면 부탁드리겠습니다...🙇🏻‍♂️

p.s. 현재 페이 구현체에서는 주문 객체를 받아 해당 객체에 적혀있는 가격을 이용하여 결제를 진행하고 있습니다.

Taewoongjung replied:
오..주문 클래스에게 책임이 너무 몰린다.. 충분히 고민 해봐야하는 사항이네요 ~!

저는 주문생성하고, 만들어진 주문 객체를 pay에 전달해주면 될거라고 생각했습니다.

Bue-von-hon replied:
🤔주문 금액에 쿠폰할인이 적용된 "최종금액"은 어디서 계산해야 할까요?

이해를 위해 정리하자면 현재 다음과 같은 방식으로 동작하고 있어요
1. 외부 결제 시스템에 "주문" 객체를 넘겨준다 (ExternalPayAPI)
2. 이 api에서 "주문" 객체를 받는 이유는 가격, 생성일 등 주문에서 필요한 정보를 받아 사용하기 때문
 
여기서 문제인점은 주문에 적혀있는 가격이 쿠폰에 의해 수정된채로 외부 결제 api에 넘어가야 하거든요
즉 우리가 들고있는 **주문과 쿠폰 도메인을 합쳐서 새로운 비즈니스 객체를 만드는 상황**입니다.

f-lab-Carrey replied:
주문 요청에 대한 기준을 가지고 도메인을 나눠보면 어떨까 싶습니다.
기본적으로 여러분들이 주문 요청을 하기 위해서는 아래의 요소들이 필요합니다.
* 주문정보 - 누가, 어디로 배달하는지, 언제 배달요청을 했는지, 필요한 요청사항은 무엇인지?
* 주문내역정보 - 무엇을 배달하는지?
* 쿠폰정보 - 어떤 쿠폰으로 금액을 할인하는지?
* 결제정보 - 어떤 결제 수단으로 결제하는지?

그렇다하면 필요한 도메인은 기본적으로 아래로 나눠볼 수 있을 것 같아요~
* 주문
  * 주문내역 - 메뉴, 수량
* 할인
* 결제 - 사실 결제는 결제라는 행위도 하지만 결제 이력도 필요할 것 같아요~ (결제 행위 자체는 DB에 저장되지 않으니)

위에 제가 적어드린 코멘트를 바탕으로 한번 더 고민해보시면 좋을 것 같아요~
의외로 도메인 간의 의존 관계를 최소화 할 수도 있을 것 같네요

Bue-von-hon replied:
@f-lab-Carrey 아... 정리하자면 이렇네요
1. 주문은 주문 정보에 대한 역할만 수행
2. 할인은 쿠폰으로 주문할인(쿠폰 객체에게 주문을 주고 "할인된 금액을 알려줘"라고 물어보는게 자연스러워 보이네요)
3. 결제는 선택한 결제수단 결제 진행

제가 이해한 바로는 아래와 같이 구현할 것 같습니다.
1. 클라이언트로부터 주문과 쿠폰을 입력 받는다.
2. 쿠폰 객체의 getDistcountPrice(Order order)를 통해 할인된 주문금액을 받는다.
3. "**외부 페이 메소드 구현체**"(여기서는 ExternalKakaoPayImpl)의 pay(int price)를 통해 결제를 호출한다
4. "**외부 페이 메소드" 구현체**"는 PayResponse를 반환한다. (요 응답도 벤더마다 다를 것 같은데... 흠...)
5. "**페이 메소드 구현체**"(e.g. kakaoPay)에서 resp를 통해 성공/실패를 확인한다.

p.s. 결제 내역은 도메인 줄일때 없앴는데 이번 이슈에서 "부활"시켜야 할까요?

f-lab-Carrey replied:
> 쿠폰 객체에게 주문을 주고 "할인된 금액을 알려줘"라고 물어보는게 자연스러워 보이네요

이거는 굳이 주문 전체를 던질 필요가 있을지도 한번 고민 부탁드려요~ 주문을 던지는 순간 쿠폰 -> 주문에 의존되게 됩니다.
쿠폰 쪽에서 필요한 데이터만 주문에서 뽑아서 요청하는 관점은 어떨까요~?

> 결제 내역은 도메인 줄일때 없앴는데 이번 이슈에서 "부활"시켜야 할까요?

현 시점에서는 굳이 부활시키지 않아도 좋을 것 같습니다~!

Bue-von-hon commented on src/main/java/app/foodylab/adaptor/payment/ExternalKakaoPayImpl.java:
```diff
@@ -1,28 +1,33 @@
-        }
-        approvePay();
-        return true;
+    public boolean processPay(Order order) {
```
🤔여기 리턴 타입이 Resp여야 할 것 같네요
Impl 클래스는 저희가 구현한게 아니라 다른 라이브러리를 사용할테니까요

Bue-von-hon commented on src/main/java/app/foodylab/adaptor/payment/KakaoPayResponse.java:
```diff
@@ -0,0 +1,17 @@
+import lombok.RequiredArgsConstructor;
+
+@Getter
+@RequiredArgsConstructor
```
🤔접근 레벨 고민해보겠습니다..!

Bue-von-hon replied:
이 친구는 외부에서 생성할 일이 없어 보여요
그래서 패키지로 설정했습니다!
[425b538](https://github.com/f-lab-edu/foody-lab/pull/43/commits/425b53819d2e0dd559ad969a902c0b33381dcd9e)

f-lab-Carrey commented:
결제에 주문 객체를 넘기도록 결정한 배경에는 어떤 것들이 있을까요~?

f-lab-Carrey commented on src/main/java/app/foodylab/adaptor/payment/ExternalKakaoPayImpl.java:
```diff
@@ -1,28 +1,33 @@
+        return "200".equals(responseCode);
     }
 
     private boolean requestPay() {
         return true;
     }
```
이 함수는 의미가 없는 함수라고 생각이 되는데요~ 추후에 어떤 용도로 발전시킬 계획이신지 궁금합니다

Bue-von-hon replied:
결제가 준비되었다는 요청을 보내는 기능으로 고려했습니다!
결제 요청 -> 결제 승인 이런식으로요!

Bue-von-hon commented:
> 결제에 주문 객체를 넘기도록 결정한 배경에는 어떤 것들이 있을까요~?

카카오api를 살펴보니 주문이 생성된 날짜, 주문의 금액, 주문에 포함된 메뉴(우리는 없습니다..) 등이 포함되어 있더라구요
그래서 주문에서 가격 속성만 넘겨주는게 아니라 주문 객체를 넘겨주어야겠다고 생각했습니다..!

f-lab-Carrey commented on src/main/java/app/foodylab/domain/order/Order.java:
```diff
@@ -9,13 +9,13 @@
     private final LocalDate orderedDate;
 
-    public static Order of(long id, String customerId, String storeId, long orderAmount,
+    public static Order of(Long id, String customerId, String storeId, long orderAmount,
```
기본 타입을 쓰는 연습을 해보면 좋겠습니다~
또 id의 경우는 DB에 의존하여 채번이 이루어지게 됩니다. 도메인 객체를 생성할 때 id를 입력해야할 필요가 있을까요?

Bue-von-hon replied:
흠... 이 친구는 디비에 저장이 되지 않는걸로 보이네요(주문 내역이 저장된다 해도 "주문내역"클래스가 있을테니...)
불필요한 속성은 제거하겠습니다!

Bue-von-hon replied:
흠... 여기를 고치니까 테스트가 깨지더라구요
깨진 테스트 고치니까 테스트에 "냄새"나는 부분이 보였어요(중복이 많더라구요)
중복도 제거하니까 커밋이 커졌네요...
[d0f089e](https://github.com/f-lab-edu/foody-lab/pull/43/commits/d0f089e0059e5d950f7e8269a1c7a2f130eb797c)

f-lab-Carrey replied:
id의 경우엔 생성 시에는 모르다가 저장 이후 조회 할 때 처음으로 등장하는 값으로 생각됩니다~
이 관점에 맞게 생성자 선언을 부탁드려요~

Bue-von-hon replied:
of 메소드를 하나 더 만들었습니다
하나는 id없이, 다른 하나는 id를 받아서 구현했는데요
id가 없으면 default 값으로 -1L을 넣는게 마음에 걸리네요(그냘 레퍼클래스로 바꿔서 null을..?)
[18782b1](https://github.com/f-lab-edu/foody-lab/pull/43/commits/18782b15acb15cb48dd839bec2e8ff9e58e610c4)

edit kakao api를 살펴보니 주문아이디가 String 타입이더라구... 일단 string으로 구현중입니다...

f-lab-Carrey replied:
JPA의 경우를 예로 들면 JPA의 경우엔 아래와 같은 규칙으로 insert하는 데이터인지 update하는 데이터인지 판별합니다.
* ID의 타입이 Number 타입의 구현체 (int, long..)인 경우 
  * id = 0이면 새로 생성하는 객체로 취급하여 insert 합니다.
  * id != 0이면 이미 등록된 객체로 취급하여 update 합니다. (만약 DB에 없다면 NotFoundException이 발생합니다)
* ID 타입이 레퍼런스 타입인 경우 
  * id = null이면 새로 생성하는 객체로 취급하여 insert 합니다.
  * id != null이면 이미 등록된 객체로 취급하여 update 합니다.

위의 관점을 고려해보시는 것도 좋을 것 같네요~

Taewoongjung replied:
오.. 이런 기능이 있었군요!
파라메터에 Long id 가 있는 of 메소드, long id 가 있는 of 메소드가 있어도 될 것 같네요!

f-lab-Carrey commented on src/main/java/app/foodylab/domain/payment/DiscountPriceCalculateService.java:
```diff
@@ -7,15 +7,16 @@
     }
 
-    public long getPrice(Order order, List<Coupon> coupons) {
+    public Order getPrice(Order order, List<Coupon> coupons) {
```
리턴타입이 메소드 명과 맞지 않네요

Bue-von-hon replied:
수정했습니다!
[f4354f3](https://github.com/f-lab-edu/foody-lab/pull/43/commits/f4354f3e2cfbd667ee39ff26d2fa5cb9658b123b)

Bue-von-hon committed:
명확한 이름으로 변경

Bue-von-hon committed:
접근 제어레벨 패키지로 설정

Bue-von-hon committed:
Order 클래스의 id제거 및 테스트 수정

order 클래스에서 id를 제거했습니다
테스트 코드를 수정하여 가독성을 향상시켰습니다
OrderFixture의 인스턴스 객체 이름을 dummy로 수정했습니다

f-lab-Carrey commented:
> 카카오api를 살펴보니 주문이 생성된 날짜, 주문의 금액, 주문에 포함된 메뉴(우리는 없습니다..) 등이 포함되어 있더라구요
그래서 주문에서 가격 속성만 넘겨주는게 아니라 주문 객체를 넘겨주어야겠다고 생각했습니다..!

결제를 하는데 필요한 속성이 주문 전체의 데이터일까요?
필요한 데이터를 받고 결제라는 도메인에서 필요한 request 스펙을 선언하고 맞춰주는 건 어떨까요?

```java
public boolean pay(PayRequest request) {
  // todo 결제 요청
}

payService.pay(createPayRequest(order)) // order에서 정보를 뽑아 결제 요청에 필요한 객체를 생성하기
```

Bue-von-hon committed:
Order 클래스의 id추가

id를 다시 추가하고, of 메소드를 추가했습니다

Bue-von-hon committed:
of 메소드 재사용하여 반복되는 코드 삭제

Bue-von-hon committed:
쿠폰에게 가격만 넘겨주고 할인된 가격을 가져옵니다

Bue-von-hon committed:
결제 진행 과정을 대대적으로 수정했습니다

ready, approve가 단계적으로 진행됩니다.
각각 request, response가 필요하고 생성됩니다.
ready 단계에서 받은 응답을 바탕으로 승인 요청을 보내게 됩니다.

Bue-von-hon committed:
refactor PayReadyRequest

url은 이미 형식이 있고, 뒤에 붙는 토큰만 다르게 설정할 수 있도록 변경

Bue-von-hon committed:
테스트 깨진거 수정

Bue-von-hon committed:
refactor PayReadyRequest

정적 팩토리 메소드를 활용해 파라미터 크기 줄임
set메소드의 접근을 private으로 변경

Bue-von-hon committed:
polishing

Bue-von-hon committed:
refactor KakaoPay

카카오 페이 클래스 가독성 향상시켰습니다
승인요청 클래스를 정적 팩토리로 생성하게 수정했습니다

Bue-von-hon committed:
polishing

Bue-von-hon committed:
주석 추가

Bue-von-hon commented:
@f-lab-Carrey @Taewoongjung 
카카오 페이 api를 뜯어보면서 싹 고쳤습니다..(제가 크게 오해했네요..)
결제를 진행하기 위해서는 크게 2가지 단계가 필요했습니다 (ready, approve)
각각 단계에 필요한 request, response를 제작했습니다..!

쿠폰도 여러개가 아닌 하나만 받도록 수정했습니다, 이 부분은 고민중입니다.
하나만 사용하면 구현은 편하지만 제공하는 서비스 퀄리티는  떨어집니다.
여러개를 정책에 맞게 사용할수 있으려면 구현은 복잡해집니다.

아래는 클래스 다이어그램과 시퀀스 다이어그램입니다.(모두 인텔리제이가 만들어줬습니다)
![payment](https://user-images.githubusercontent.com/46879264/200126095-d11a26af-2f0e-44cb-bdd6-33d83fdc92f6.png)
![KakaoPay_pay](https://user-images.githubusercontent.com/46879264/200151994-54f7bf8d-9dbe-4248-9078-7f219514c5c5.png)


카카오 페이 클래스에서 ExternalKakaoPayImpl 이 친구의 ready, approve를 호출하면서 결제를 진행시킵니다.
ExternalKakaoPayImpl 이 친구는 아예 외부 라이브러리라고 생각하면 될 것같습니다
p.s. 수정이 꽤 많았네요... 리뷰가 힘드시겠지만 잘 부탁드리겠습니다...🙇🏻‍♂️🙇🏻‍♂️

f-lab-Carrey commented on src/main/java/app/foodylab/adaptor/payment/ExternalKakaoPayImpl.java:
```diff
@@ -1,28 +1,23 @@
-        }
+    @Override
+    public PayApproveResponse approvePay(PayApproveRequest request) {
+        // todo: requestId 요청고유번호의 역할은..?
```
외부 API에서 주는 requestId는 말 그대로 요청에 대한 식별자입니다.
추후에 카카오페이에 요청 했던 결제 정보를 조회할 때 보통 requestId를 기준으로 조회하게 됩니다~

Bue-von-hon replied:
이해했습니다!

f-lab-Carrey commented on src/main/java/app/foodylab/application/payment/PayMethod.java:
```diff
@@ -1,8 +1,11 @@
     String getMethod();
 
-    boolean pay(long amount);
+    PayApproveResponse pay(PayReadyRequest request);
```
PayReadyRequest와 PayApproveResponse는 카카오페이의 스펙으로 생각이 됩니다.
만약 네이버페이를 추가한다고 할 경우 도메인 레벨에서 똑같은 요청 스펙을 유지할 수 있을까요?

Bue-von-hon replied:
🤔이 부분은 외부와 연결된 부분이었네요
그렇다면 인터페이스를 통해 구현해야할 것 같아요
그리고 아마도 payResponse도 벤더사마다 다를것 같은데요..?
pay응답을 (카카오, 네이버, 토스) 중 하나, 이런 식으로 조합해보겠습니다

f-lab-Carrey replied:
결제 요청을 위한 공통된 스펙들이 있을 것이라 생각됩니다~ 공통적인 부분만 추출해보면 추상화를 해볼 수 있을 것 같아요

Taewoongjung replied:
[네이버] (https://developer.pay.naver.com/docs/v2/api)
[토스](https://tossdev.github.io/api.html#httpResponse)
[카카오] (https://developers.kakao.com/docs/latest/ko/kakaopay/single-payment)


공통적으로 들어가는 파라미터
- 가맹점 id
- 총 결제 금액
- 상품명
- 상품 비과세 금액
- 사용자 결제 완료 후 결제 결과를 받을 URL
등등이 있겠네요

참고해주세요^^

f-lab-Carrey commented on src/main/java/app/foodylab/domain/order/Order.java:
```diff
@@ -9,19 +9,24 @@
     private final LocalDate orderedDate;
 
-    public static Order of(long id, String customerId, String storeId, long orderAmount,
+    public static Order of(String customerId, String storeId, long orderAmount,
+        LocalDate orderedDate) {
+        return of(null, customerId, storeId, orderAmount, orderedDate);
+    }
+
+    public static Order of(String id, String customerId, String storeId, long orderAmount,
```
of 메서드가 두개가 될 이유가 있을까요? 두번째 of 메서드는 실상 첫번째 of에서만 호출 되는 것 같아서요~

Bue-von-hon replied:
id가 필수가 아니라 생각되어 final을 사용하지 않았거든요
그러다보니 이런식으로 바리에이션을 주려했던것 같네요... 괜히 더 복잡해지는것 같아서 삭제했습니다!
(테스트를 위한 코드를 만든 느낌도 드네요..ㅎ 프로덕션 코드에서는 딱 한곳에서 쓰이고 있긴 했지만요...)
[2aa53fc](https://github.com/f-lab-edu/foody-lab/pull/43/commits/2aa53fc452fa2386bbed4d76f95a5fe1b0f6d2e4)

f-lab-Carrey commented on src/main/java/app/foodylab/domain/payment/PayReadyRequest.java:
```diff
@@ -0,0 +1,32 @@
+    private final long price;
+    private final String storeId;
+    private final String orderId;
+    private String approvelUrl = "/approve";
+    private String cancelUrl = "/cancel";
+    private String failUrl = "/fail";
```
이 url들은 어떤 용도 일까요?

Bue-von-hon replied:
카카오 페이로 결제 준비요청을 보낼때 "성공하면 여기로, 취소되면 여기로, 실패하면 여기로 redirect해줘"라는 내용을 보내더라구요
MVP 스펙으로 구현중이라 하나만 있어도 될 것 같았지만 이왕 구현하는 김에 카카오페이 스펙 그대로 구현했습니다..!
[링크](https://developers.kakao.com/docs/latest/ko/kakaopay/single-payment#prepare)

f-lab-Carrey commented:
좋은 방향으로 발전이 되고 있다고 생각합니다~ 남겨드린 코멘트를 확인 부탁드리겠습니다

Bue-von-hon committed:
refactor : 불필요한 of 메소드 삭제

f-lab-Carrey commented:
conflict 해결 후에 1차적으로 merge를 하면 좋겠습니다 
이 PR은 호흡이 조금 길어서 작업자가 지칠수도 있을 것 같네요~

Bue-von-hon committed:
Merge remote-tracking branch 'origin/dev' into issue-37-KakaoPayImpl_리팩터링

# Conflicts:
#	src/main/java/app/foodylab/application/coupon/Coupon.java
#	src/main/java/app/foodylab/application/payment/service/PayService.java
#	src/main/java/app/foodylab/domain/coupon/PriceCoupon.java
#	src/test/java/app/foodylab/application/payment/service/PayServiceTest.java

Bue-von-hon committed:
conflict 해결 및 깨진 테스트 수정

Bue-von-hon committed:
refactor : 변수이름 수정, 반복되는 테스트 파라미터화

Bue-von-hon committed:
refactor : import문 정리, 불필요한 코드 삭제

Bue-von-hon commented:
> conflict 해결 후에 1차적으로 merge를 하면 좋겠습니다 이 PR은 호흡이 조금 길어서 작업자가 지칠수도 있을 것 같네요~

넵 충돌해결했고, 새로운 이슈에서 남겨둔 리뷰를 반영해보겠습니다~!
